### PR TITLE
deepCopy Functionality

### PR DESCRIPTION
**Pull request by ThinkTankShark**
Extracts and improves the deepCopy, adds tests

### CODE REVIEW

File: spec/lib/deepCopy_spec.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    var result = deepCopy(arr);
+
+    expect(result.length).toEqual(arr.length);
+    expect(result[0].config.port).toBe(arr[0].config.port);
+    expect(result[0].config.host).toBe(arr[0].config.host);
+    expect(result[1].config.deepConfig.db).toBe(arr[1].config.deepConfig.db);
+    expect(result[1].config.deepConfig.ssl).toBe(arr[1].config.deepConfig.ssl);
+    expect(result[1].config.cache.type).toBe(arr[1].config.cache.type);
+  });
+});
```
Comment (isuttell) [inline spec/lib/deepCopy_spec.js]: Let's add another spec to test mutations. Mutation where the reason I created this method

File: src/lib/deepCopy.js [hunk 2]
```diff
@@ -5,14 +5,39 @@
...
  */
 module.exports = function deepCopy(obj) {
+  var copy;
+
   if (obj === null || typeof obj !== 'object') {
     return obj;
   }
-  var src = obj instanceof Array ? [] : {};
-  for (var key in obj) {
-    if (Object.prototype.hasOwnProperty.call(obj, key)) {
```
Comment (isuttell) [inline src/lib/deepCopy.js]: We need this line. At some point I ran across another external library that nuked `hasOwnProperty` so I had to call it directly off of the Object.prototype.

File: spec/lib/deepCopy_spec.js [hunk 3]
```diff
@@ -103,4 +113,58 @@ describe('deepCopy', function(){
...
+    expect(result).not.toBe(tree);
+    expect(result.left).toEqual(tree.left);
+    expect(result.right).toEqual(tree.right);
+
+    // Now changing the source object,
+    // but result should not be changed
+    tree.right = tree.left;
+
+    // There are not a same object in memory
+    expect(result).not.toBe(tree);
```
Comment (isuttell) [inline spec/lib/deepCopy_spec.js]: toBe compares using `===` so it's a dup from above

File: spec/lib/deepCopy_spec.js [hunk 4]
```diff
@@ -103,4 +113,58 @@ describe('deepCopy', function(){
...
+    expect(result.left).toEqual(tree.left);
+    expect(result.right).toEqual(null);
+    expect(result.right).not.toEqual(tree.right);
+  });
+
+  it('should copy the object correctly, not mutate', function() {
```
Comment (isuttell) [inline spec/lib/deepCopy_spec.js]: What is the difference in this test compared to the above?

File: src/lib/deepCopy.js (hunk 2 above)
Comment (ThinkTankShark) [inline src/lib/deepCopy.js]: Fixed.

File: spec/lib/deepCopy_spec.js (hunk 4 above)
Comment (ThinkTankShark) [inline spec/lib/deepCopy_spec.js]: Fixed and optimized the tests

File: spec/lib/deepCopy_spec.js (hunk 3 above)
Comment (ThinkTankShark) [inline spec/lib/deepCopy_spec.js]: Removed the excessive expects

File: spec/lib/deepCopy_spec.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    expect(result.first).toEqual(obj.first);
+
+    // Now changing the source object,
+    // but result object should not be changed
+    obj.second = false;
+
+    expect(result.second).not.toEqual(obj.second);
+  });
+
+  it('should copy the object correctly, not mutate', function() {
```
Comment (isuttell) [inline spec/lib/deepCopy_spec.js]: This is the same as the one above it right?

Comment (isuttell) [inline spec/lib/deepCopy_spec.js]: Let's just leave it as one

**Pull request merged**